book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+cluster. This means either (planned) hard-fork, urgent security fix requiring
+ABI change (Tier-1), or normal periodical voluntary software update (Tier-2).
+Tier-1 includes transactions and accounts, for example. Tier-2 includes shreds
+and snapshots, for example.
+
+Hard fork: A hard fork is a change to the cluster that makes previously invalid
+communications valid, and therefore requires all users / software to upgrade
+(definition is based on [1]). The upgrade is programmed and scheduled and takes
+place effective at specific future epoch.
+
+# Examples
+
+## Example 1: If we were adding #7233 after some time since mainnet launch:
+
+```patch
+@@applicaton code@@
+
+ #[solana_abi(allow=add, digest="79caeee9")]
+   // ^^ adding a member to allow=add enum doesn't trigger digest change
+ #[derive(SolanaABI)]
+ pub enum VoteError {
+     ....
++
++    #[error("vote timestamp not recent")]
++    TimestampTooOld,
+ }
+
+-#[solana_abi(allow=none, digest="3ffc9921")]
++#[solana_abi(allow=none, digest="1c6a53e9")]
+ #[derive(Serialize, Default, Deserialize, Debug, PartialEq, Eq, Clone)] 
+ pub struct Vote {
+     /// A stack of votes starting with the oldest vote
+     pub slots: Vec<Slot>,
+     /// signature of the bank's state at the last slot
+     pub hash: Hash,
++    #[solana_abi_added(since = "rev1")]
++    /// processing timestamp of last slot
++    pub timestamp: Option<UnixTimestamp>,
+ }
+
+ impl SolanaABI for Vote {
+   const REVISION_LIST = [
+      (0, "epoch", 0..), // leave this revision acceptable for indefinite time
++     (1, "epoch", 21..),
+   ]
+ }
+
+
+@@test code@@
+ #[test]
+ #[solana_abi_test]
+ test_serialize_vote() {
+    //   ^^^ no change is needed for ABI test case
+    let vote = ...
+    serializing_with_fixture!(vote, |write_stream, revision_tuple| {
+        // use revision_tuple to differentiate serialization in some way...
+        serialize_into(write_stream, normal_vote)
+    });
+    ...
+ }
+
+ #[test]
+ #[solana_abi_test]
+ test_deserialize_vote() {
+    //   ^^^ no change is needed for ABI test case
+    let vote = ...
+    deserializing_with_fixture!(Ok(vote), |read_stream, revision_tuple| {
+        // use revision_tuple to differentiate deserialization in some way..
+        deserialize_from(write_stream)
+    });
+ }
+
+Files:
+ test/fixtures/vote-rev-0.hex.yaml
++test/fixtures/vote-rev-1.hex.yaml
+```
+
+## Example 2: If we were adding #7415 after some time since mainnet launch:
+
+```patch
+ /// Meta contains enough context to recover the index from storage itself
+ /// This struct will be backed by mmaped and snapshotted data files.
+ /// So the data layout must be stable and consistent across the entire cluster!
+-#[solana_abi(allow=none, digest="662792c9", test="test_hash_stored_account")]
++#[solana_abi(allow=none, digest="c5031113", test="test_hash_stored_account")]
+ #[derive(PartialEq, Debug)]
+ pub struct StoredAccount<'a> {
```

Comment:
Taking `StoredAccount` as an example here.   I'd prefer that once it's first given a `solana_abi` attribute and digest, it may generally never modified again once the software that first contains its definition ships.   The main exception to this would be if we decide we don't care to retain backwards compatibility for some future release (a deliberate hard fork).

Should a field need to be added or removed, we need to create an entirely new `StoredAccount1` struct (followed by `StoredAccount2`, etc).  At runtime both types will likely be needed.  For example, the old `StoredAccount` is needed when processing older ledger data or dealing with validator running older software.

Reply:
Thinks for clarification, I've added [a paragraph]https://github.com/solana-labs/solana/pull/7524/files#diff-7bbc1dbd50f4144bb25bfb9d3c28257aR74-R77() accordingly!